packages/ember-metal/lib/expand_properties.js

```diff
@@ -40,55 +41,34 @@ export default function expandProperties(pattern, callback) {
     'Brace expanded properties cannot contain spaces, e.g. "user.{firstName, lastName}" should be "user.{firstName,lastName}"',
     pattern.indexOf(' ') === -1
   );
+  assert(
+    `Brace expanded properties have to be balanced and cannot be nested, pattern: ${pattern}`,
+    pattern.match( NESTED_BRACES ) === null
+  );
 
-  let unbalancedNestedError = `Brace expanded properties have to be balanced and cannot be nested, pattern: ${pattern}`;
-  let properties = [pattern];
+  if (pattern.indexOf('{') < 0) {
+    callback( pattern.replace(END_WITH_EACH_REGEX, '.[]') );
+  } else {
+    dive("", pattern, callback);
+  }
+}
 
-  // Iterating backward over the pattern makes dealing with indices easier.
-  let bookmark;
-  let inside = false;
-  for (let i = pattern.length; i > 0; --i) {
-    let current = pattern[i - 1];
+function dive(prefix, pattern, callback) {
+  let start = pattern.indexOf('{');
```

Comment:
yeah that optimization was for the most likely case of no-expansion.  I did look at re-using once already but I'll look again with some freshly caffeinated eyes :D